lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,270 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <optional>
+
+#include "mlir/Bytecode/BytecodeReader.h"
+#include "mlir/Bytecode/BytecodeWriter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/FunctionInterfaces.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Parser/Parser.h"
+
+#include "revng/Pipeline/RegisterContainerFactory.h"
+
+#include "revng-c/Pipes/Kinds.h"
+#include "revng-c/mlir/Pipes/MLIRContainer.h"
+
+#include "MLIRLLVMHelpers.h"
+
+using namespace revng;
+using namespace revng::pipes;
+using namespace revng::mlir_llvm;
+
+static constexpr auto &TheKind = kinds::MLIRLLVMFunctionKind;
+
+using ContextPtr = std::unique_ptr<mlir::MLIRContext>;
+using OwningModuleRef = mlir::OwningOpRef<mlir::ModuleOp>;
+
+static OwningModuleRef cloneModuleInto(mlir::ModuleOp SourceModule,
+                                       mlir::MLIRContext &DestinationContext) {
+  llvm::SmallString<1024> Buffer;
+  llvm::raw_svector_ostream Out(Buffer);
+  mlir::writeBytecodeToFile(SourceModule, Out);
+
+  mlir::Block OuterBlock;
+  mlir::LogicalResult
+    R = mlir::readBytecodeFile(llvm::MemoryBufferRef(Buffer, "cloneModuleInto"),
+                               &OuterBlock,
+                               mlir::ParserConfig(&DestinationContext));
+  revng_assert(mlir::succeeded(R));
+
+  auto &Operations = OuterBlock.getOperations();
+  revng_assert(Operations.size() == 1);
+
+  auto DestinationModule = mlir::cast<mlir::ModuleOp>(Operations.front());
+  DestinationModule->remove();
```

Comment:
Add a comment like "Ah we can't directly deserialize the `ModuleOp`, so we deserialize it into a block, then detach it and return it passing ownership to the caller".